protocols/syslog/syslog.c

```diff
@@ -23,146 +24,82 @@
 
 #include <avr/pgmspace.h>
 #include <stdarg.h>
+#include <stdlib.h>
 
 #include "protocols/uip/uip.h"
 #include "config.h"
 #include "core/debug.h"
-#include "protocols/uip/uip_neighbor.h"
+#include "core/queue/queue.h"
 #include "protocols/uip/uip_router.h"
 #include "protocols/uip/check_cache.h"
 #include "syslog.h"
 #include "syslog_net.h"
 
+#define UIP_MAX_LENGTH (UIP_BUFSIZE - UIP_IPUDPH_LEN - UIP_LLH_LEN)
 
-static char send_buffer[MAX_DYNAMIC_SYSLOG_BUFFER + 1];
 extern uip_udp_conn_t *syslog_conn;
-static struct SyslogCallbackCtx syslog_callbacks[SYSLOG_CALLBACKS];
+static Queue syslog_queue = { NULL, NULL };
 
-static void syslog_send_cb_P(void *data) 
-{
-  strcpy_P(uip_appdata, data);
-  uip_udp_send(strlen_P(data));
-}
-
-static void syslog_send_cb(void *data) 
-{
-  char *p = data;
-
-  strcpy(uip_appdata, p);
-  uip_udp_send(strlen(p));
-
-  if (p == send_buffer)
-    p[0] = 0;
-}
-
-
-uint8_t 
-syslog_send_P(PGM_P message)
-{
-  return syslog_insert_callback(syslog_send_cb_P, (void *)message);
-}
-
-uint8_t 
+uint8_t
 syslog_send(const char *message)
 {
-  uint16_t offset = strlen (send_buffer);
+  char *data = malloc(strlen(message) + 1);
 
-  if (strlen (message) + offset + 1 > MAX_DYNAMIC_SYSLOG_BUFFER)
+  if (data == NULL)
     return 0;
+  strncpy(data, message, strlen(message) + 1);
```

Comment:
strncpy will only copy n bytes including terminating null char. Because data is allocated at this size i see no problems with this line. If i remove the +1 the null char is missing and has to added manually. If null char is missing in message then strlen would read until it hits next null in memory and data is allocated with this size.